### ISSUE TITLE
Video Quality slider setting does not apply to "nvenc" encoder output

### ISSUE THREAD

**Issue by CPioGH2002** [C0untFloyd/roop-unleashed#222]
**Describe the bug**
When setting up the Video Quality in the webui via Settings >> "Video Quality (crf)" slider, the setting does not affect the output from using "nvenc"-based encoding. On the (CPU-based) libx264 and libx265 encoders it seems to work as intended though.

**To Reproduce** _(only applies to Nvidia users)_
Steps to reproduce the behaviour:
1. Pick "hevc_nvenc" or "h264_nvenc" in Settings >> Video Codec
2. For "Video Quality (crf)", set up a low value (=high quality) like "5"
3. Generate video, observe file size and subjective video quality
4. New run: for "Video Quality (crf)", set up a high value (=low quality) like "95"
5. Repeat step 3
6. Compare the two videos. If those are the same file size, bitrate and quality, the slider position didn't matter

**Details**
What OS are you using?
-Linux (Ubuntu 23.04)

Are you using a GPU?
-NVIDIA

**Which version of roop unleashed are you using?**
3.0.8

**Screenshots**
Imagine two identical screenshots to prove my point. :-p

**Additional info:**
To my (limited) understanding, the video quality related parameters for FFmpeg currently being used ("-crf") do not apply to nvenc based encoding, hence the lack of impact of mentioned slider setting when having nvenc in use. I have to check which parameters nvenc would need to implement a quality adjustment in conjunction with a constant bitrate. Will post if I find that info and be grateful for your suggestions of course.

**Workarounds:**
None so far. :-( Nvenc encoded videos come out with a fixed, rather low-ish setting.

**Related threads:**
There were some on video quality in general but those related to an already fixed (thumbs up for that one!) bug. Not specifically to the "Video Quality" slider in conjunction with nvenc.

**Comment by C0untFloyd** [comment]
tldr; perhaps I need to integrate a preset/bitrate advanced setting, I've seen that being used in other roop forks.

Smarty-pants Google Bard says this:

To get the best video quality when using FFmpeg with the NVENC codec, you can follow these recommendations:

* Use a constant bitrate (CBR) instead of variable bitrate (VBR). This will ensure that your video has a consistent quality throughout, even if there are complex scenes or high-motion footage.
* Use a high bitrate. The higher the bitrate, the better the quality of your video will be. However, it is important to choose a bitrate that is appropriate for your target audience and distribution platform.
* Use the `veryslow` preset. This will encode your video at the highest possible quality, but it will also take the longest. If you need to encode your video faster, you can use a faster preset, such as `slow` or `medium`, but this will result in a lower quality video.
* Use the `-rc-lookahead N` option to specify the number of lookahead frames. This will help the encoder to make better decisions about how to allocate bits, resulting in better quality video. A lookahead of 10-20 frames is suggested for optimal quality benefits.

Here is an example of an FFmpeg command that you can use to encode a video with the NVENC codec and the highest possible quality:

```
ffmpeg -i input.mp4 -c:v h264_nvenc -preset veryslow -rc-lookahead 20 -b:v 20M output.mp4
```

This command will encode the input video file `input.mp4` to the output video file `output.mp4` using the NVENC codec with the `veryslow` preset, a lookahead of 20 frames, and a bitrate of 20 Mbps.

You can adjust the bitrate and preset options depending on your specific needs. For example, if you need to encode your video faster, you can use a faster preset, such as `slow` or `medium`. Or, if you need to reduce the file size of your video, you can use a lower bitrate.

It is important to note that NVENC is a hardware encoder, so the quality of your encoded video will also depend on the performance of your NVIDIA GPU. If you have an older or less powerful NVIDIA GPU, you may not be able to achieve the same quality as with a newer or more powerful GPU.

Not so smart Meta LLama told something about fixed bitrate and presets and lying chatgpt 3.5 said this:

Hello! I'm happy to help you with your question. To get the best video quality when using the NVENC codec with FFmpeg, there are a few things you can do:

1. Make sure your NVIDIA GPU is compatible with NVENC: Check the NVIDIA website to see if your GPU supports NVENC. You can also use the `nvenc-utils` package to check if your GPU supports NVENC.
2. Install the correct NVENC driver: Make sure you have the latest NVENC driver installed. You can check the NVIDIA website for the latest drivers.
3. Use the `-c:v nvenc` flag: When encoding your video, use the `-c:v nvenc` flag to specify that you want to use the NVENC codec. For example: `ffmpeg -i input.mp4 -c:v nvenc -b:v 5000k output.mp4`
4. Adjust the encoding settings: You can adjust the encoding settings to optimize the video quality. Some options you might want to consider include:
        * Bitrate: Increase the bitrate to improve video quality. For example: `ffmpeg -i input.mp4 -c:v nvenc -b:v 5000k -crf 18 output.mp4`
        * Resolution: Set the resolution to the highest possible value for your input video. For example: `ffmpeg -i input.mp4 -c:v nvenc -s 1920x1080 output.mp4`
        * Frames per second: Set the frames per second to the highest possible value for your input video. For example: `ffmpeg -i input.mp4 -c:v nvenc -r 60 output.mp4`
5. Use a high-quality preset: You can use a high-quality preset to optimize the video quality. For example: `ffmpeg -i input.mp4 -c:v nvenc -preset ultrafast -b:v 5000k output.mp4`
6. Check the encoding logs: After the encoding process is complete, you can check the encoding

After querying the suspicious preset:

You are correct, the ultrafast preset in ffmpeg's nvenc codec is designed for faster encoding speed rather than optimal video quality. If you prioritize video quality, it is recommended to use slower presets such as slow or slower.

The ultrafast preset sacrifices encoding efficiency to achieve faster encoding times, resulting in lower video quality compared to slower presets. While it may be useful in certain scenarios where speed is crucial, it is not ideal for maximizing video quality.

To get the best video quality, you should consider using the slow or slower presets with the nvenc codec, as mentioned in the previous response. These presets prioritize encoding quality over speed, resulting in better video output.
